src/simplification/utils/circuits_db.hpp

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "src/common/csat_types.hpp"
+#include "src/utility/converters.hpp"
+
+namespace csat::simplification
+{
+
+/**
+ * Structure for storing a circuit database.
+ */
+struct CircuitDB
+{
+    std::map<std::vector<int32_t>, int32_t> subcircuit_pattern_to_index;
+    std::vector<GateIdContainer> subcircuit_outputs;
+    std::vector<std::vector<GateIdContainer>> gates_operands;
+    std::vector<int32_t> OPER_number;
+    std::vector<std::vector<GateType>> gates_operations;
+
+    /**
+     * Reads a database for simplification in a specific format.
+     * @param db_path -- path to the database text file
+     * @param basis -- the database basis in which it will be read
+     */
+    CircuitDB(std::filesystem::path const& db_path, Basis basis)
+    {
+        if (basis != Basis::BENCH and basis != Basis::AIG)
+        {
+            std::cerr << "Incorrect basis! Choose one of [AIG, BENCH]" << std::endl;
+            std::abort();
+        }
+
+        read_db(db_path);
+    }
+
+    /**
+     * Reads a database in a BENCH format.
+     * Each row of the database must encode a circuit. Where:
+     *  -- the first number in the line is the number of inputs, which are numbered from 0 to this number - 1;
+     *  -- the second number is the number of outputs;
+     *  -- next "outputs" number of numbers are output codes, which are truth tables written in decimal form;
+     *  -- next "outputs" number of numbers are indexes of outputs;
```

Comment:
Not important note: "number of numbers" sounds wierd. 

Maybe: 
"The subsequent "outputs" entries represent output codes"